Generalize domain initializer

nilsvu opened the pull request:
## Proposed changes

Moves the domain initialiser to DiscontinuousGalerkin and removes dependencies on Evolution (#1607 removes the remaining dependency). Also adds a test.

@nilsdeppe we talked about having this in Domain instead, but neither of us liked that very much. I placed this in DG/Actions now, since it's to set up a DG simulation, and for other kinds of simulations one may want to set up the domain differently.

### Types of changes:

- [ ] Bugfix
- [ ] New feature

### Component:

- [ ] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

nilsvu committed:
Move domain initializer to DiscontinuousGalerkin

nilsvu committed:
Move tag for initial extents to Domain

Also removes Initialization::Tags::Domain because we can use
::Tags::Domain instead.

nilsvu commented on src/Evolution/Initialization/Evolution.hpp:
```diff
@@ -70,11 +70,12 @@ struct Evolution {
+        ::Tags::DerivCompute<
+            variables_tag,
+            ::Tags::InverseJacobian<::Tags::ElementMap<dim>,
+                                    ::Tags::Coordinates<dim, Frame::Logical>>,
```
FYI, this change is needed because I switched the Jacobian compute item to the `Tags::Coordinates<Dim, Frame::Logical>` base tag in the initializer.

kidder commented:
I thought we didn't want anything parallel in NumericalAlgorithms anymore and were putting these type of things in ParallelAlgorithms...

nilsvu commented:
@kidder would you prefer I put this directly in ParallelAlgorithms/DG? I thought we're going to move the whole NumericalAlgorithms/DG directory at once.

kidder commented:
I would much rather move it once than twice...   I would move it to `ParallelAlgorithms/Initialization/Actions`   thoughts @nilsdeppe ?

nilsdeppe commented:
Yea, moving once instead of twice sounds good. The location also sounds good to me. We will need some distinction for initialization different algorithms, e.g. DG vs. DG-FD vs FD vs DG-FV, etc. but I think that's just a file and action rename later if it's necessary.

nilsvu commented:
We don't want to put initializers for various parts of the code all in `ParallelAlgorithms/Initialization` though, right? Then that directory basically depends on every other part of the code. I'd rather keep the initialize actions with the other actions for a given module, for instance have the limiter initializer be in the Limiters/Actions directory. Therefore I suggest to have this domain initializer either in `Domain/Actions` (if it's general enough, which it probably isn't) or in `ParallelAlgorithms/DiscontinuousGalerkin/Actions/InitializeDomain.hpp`. When we add actions for non-DG schemes, they can go in separate directories in `ParallelAlgorithms`, such as `PA/FiniteDifference/Actions/`. Thoughts?

nilsvu commented:
@kidder @nilsdeppe could you give feedback on this thread if you get the chance? Thanks

nilsdeppe commented:
I agree that sticking everything into one directory creates a lot of dependencies... So far we've been keeping actions bundled with their libs so that seems reasonable.

nilsvu commented:
@osheamonn and @kidder could you review this perhaps (GitHub suggested you 😄)? Then we can leave @nilsdeppe working on subcell :)

As always, please feel free to forward the review to someone else if you don't have time!

nilsvu commented:
Or perhaps @moxcodes @fmahebert? I'm a bit blocked by not having the elliptic initialisation in `develop`, so I'd appreciate if anyone could help me with the review :)

osheamonn commented on src/NumericalAlgorithms/DiscontinuousGalerkin/Actions/InitializeDomain.hpp:
```diff
@@ -40,76 +40,90 @@ class ConstGlobalCache;
+ * - Modifies: nothing
+ */
+template <size_t Dim, ::Initialization::MergePolicy MergePolicy =
+                          ::Initialization::MergePolicy::Error>
```
What is `MergePolicy` and why is it being added?

nilsvu replied:
It was introduced in #1579 for `::Initialization::merge_into_databox`, which is now used in all initialisers. It allows multiple initialisers to add the same tags (see https://spectre-code.org/group__initializationgroup#gafc4ef63b38e015766414946b0b5523b5). The `MergePolicy` decides what to do with conflicts. It also makes dealing with DataBox chains in action lists easier. I suggest to generally template initializers on the `MergePolicy` and default it to `Error` so that this mechanism can be used.

nilsvu replied:
This domain initialiser is particularly tricky because `ElementMap`s are not comparable. So one can't just have an action list `tmpl::list<InitializeDomain<Dim>>` by itself: The DataBox that comes out of an iteration of this list contains a `Tags::ElementMap`, but it must be compatible with feeding it back into the start of the list. `MergePolicy::Error` fails here, because when the second iteration tries to add an `ElementMap` again it can't check if it's equal to the existing one. Using `MergePolicy::IgnoreIncomparable` would fix this.

kidder replied:
@nilsleiffischer can you give a concrete example of when one would need to use a `MergePolicy` because I really, really dislike it and strongly disagree with generally templating initializers with it...

nilsvu replied:
An example would be re-initializing the linear solver: In an executable that does multiple linear solves, one sets `MergePolicy::Overwrite` for the linear solver initializer.

If we don't template initializers on the `MergePolicy`, then there is no way to use this functionality of `merge_into_databox`, since the function will only ever be used in initialisers.

kidder replied:
I don't like this functionality in `merge_into_databox`, it seems like a hack to me.  I don't think initialization actions should be used outside of the initialization phase.  Why can't you just mutate the existing items when re-initializing the linear solver?

nilsvu replied:
I could, but using the same initializer for re-initialization seems quite elegant to me and reduces code duplication. But let's talk about the linear solver when the time comes. I really need to move forward with these PRs. I removed the `MergePolicy` here for now and instead work around the issue in the test. Please check the fixup :)

osheamonn commented on tests/Unit/NumericalAlgorithms/DiscontinuousGalerkin/Actions/Test_InitializeDomain.cpp:
```diff
@@ -0,0 +1,296 @@
+  {
+    INFO("1D");
+    // Reference element:
+    // [ |X| | ] -xi->
```
I don't really get this ascii diagram, what is it trying to convey?

nilsvu replied:
It shows the domain that is set up in the following lines. We are testing the element with index 1 at refinement level 2 (marked with `X`). How is this unclear, or do you have a suggestion how to make it clearer? Similar diagrams are used in other tests, e.g. in `Test_FluxCommunication.cpp`

osheamonn replied:
Alright, it makes more sense looking at the 2d one. Is the minus sign in front of `xi` intentional? Looking at the element ID (2nd out of 4 elements) and the map (-0.5, 1.5) -> (2, 4) doesn't suggest the `xi` axis is being flipped to me; maybe I'm missing something.

nilsvu replied:
It's not supposed to be a minus sign but part of the arrow 😄 But I see how this is misleading, will change it

nilsvu committed:
Use generalized domain initializer for elliptic executable

This includes the ExportCoordinates executables, which used
the elliptic domain initializer before.

Also note that the elliptic initializer tests are disabled for now.
They will be superseded by the upcoming generalized initializers.

nilsvu commented:
I added another commit that uses this initializer for the elliptic code. It should be quite uncontroversial, mostly deletes code :)

One thing to note is that it disables the elliptic initializer tests for now, since they rely on the old domain elliptic domain initializer and don't use the action testing framework yet. In the following PRs when generalizing the initializers one-by-one I'll also re-enable these tests one-by-one, as I do for the domain initializer in this PR.

nilsvu commented:
@osheamonn @kidder please check the fixups and let me know when to squash them.

nilsvu committed:
Generalize domain initializer

Cleans up code and documentation and adds a test.
This initializer will be used by both the evolution and elliptic code.

nilsvu committed:
Move elliptic domain initializer test to DiscontinuousGalerkin

nilsvu commented:
Thanks for reviewing @osheamonn 👍 Travis passes. This is ready for a final look @kidder or @nilsdeppe (we reviewed this together in the call yesterday)
